### PR TITLE
Adds reserved properties for filtering property definitions (second attempt)

### PR DESCRIPTION
**Pull request by pauldambra**
## Changes

Second attempt at #8291 
reverted in #8340 after causing https://sentry.io/organizations/posthog2/issues/2968014180/?project=1899813

This restricts reserved properties to events filters

## How did you test this code?

Adding developer tests that exercise the API

### CODE REVIEW

File: ee/clickhouse/models/property.py [hunk 1]
```diff
@@ -375,13 +375,14 @@ def get_property_string_expr(
     :return:
     """
 
-    for reserved_words in EVENT_ATTRIBUTE_RESERVED_PROPERTIES_BY_TYPE.values():
-        if property_name in reserved_words:
-            return property_name, False
+    table_string = f"{table_alias}." if table_alias != None else ""
 
-    materialized_columns = get_materialized_columns(table) if allow_denormalized_props else {}
+    if table == "events":
```
Comment (hazzadous) [inline ee/clickhouse/models/property.py:419]: Is reserved properties just applicable to events, or is it something we'd also want to do for other tables, but we just want to limit scope for now. Which ever way, somehow documenting would be useful

File: ee/clickhouse/views/test/test_clickhouse_trends.py [hunk 2]
```diff
@@ -357,6 +358,97 @@ def test_insight_trends_compare(self):
...
+                date_to="2020-01-02T00:00:00Z",
+                events=[{"id": "$pageview", "name": "$pageview", "type": "events", "order": 0,},],
+                properties=[{"key": "distinct_id", "value": "person1", "operator": "ncontains", "type": "person"}],
+            ),
+            team=self.team,
+        )
+        people = get_people_from_url_ok(client=self.client, url=response["result"][0]["persons_urls"][0]["url"])
+        assert {distinct_id for person in people for distinct_id in person["distinct_ids"]} == {"p1"}
+
+    def test_can_filter_trends_by_event_timestamp(self):
```
Comment (hazzadous) [inline ee/clickhouse/views/test/test_clickhouse_trends.py:396]: There isn't much context as to why we're testing timestamp here, is there something special about timestamp? Is it that it is a reserved keyword?

If it's relevant only to the regression above, perhaps we can group them under a separate testcase class for this. If it's just testing the reserved word feature maybe we can group under a feature specific testcase class. To help with navigating the tests by feature.

File: ee/clickhouse/models/property.py [hunk 3]
```diff
@@ -233,13 +233,17 @@ def prop_filter_json_extract(
         assert isinstance(prop.value, str)
         prop_value_param_key = "v{}_{}".format(prepend, idx)
 
+        if prop.key in EVENT_ATTRIBUTE_RESERVED_PROPERTIES_BY_TYPE["DateTime"]:
```
Comment (hazzadous) [inline ee/clickhouse/models/property.py:236]: Does anything need to be done here about checking prop.key != 'event'

File: ee/clickhouse/views/test/test_clickhouse_trends.py [hunk 4]
```diff
@@ -357,6 +358,97 @@ def test_insight_trends_compare(self):
...
+                properties=[
+                    {"key": "timestamp", "value": "2020-01-02 03:30:00", "operator": "is_date_after", "type": "event"}
+                ],
+            ),
+            team=self.team,
+        )
+        people = get_people_from_url_ok(client=self.client, url=response["result"][0]["persons_urls"][0]["url"])
+        assert {distinct_id for person in people for distinct_id in person["distinct_ids"]} == {"p2"}
+
+    def test_can_filter_trends_by_event_property_holding_a_datetime(self):
```
Comment (hazzadous) [inline ee/clickhouse/views/test/test_clickhouse_trends.py:422]: Is testing materialized relevant here?

**Review by hazzadous — COMMENTED** [review]
Just some small comments. I've looked at only new changes and haven't reviewed the original PR, was there anything else from that PR to review?

File: ee/clickhouse/models/property.py (hunk 1 above)
Comment (pauldambra) [reply ee/clickhouse/models/property.py:419]: Yep, limiting scope for now. But I'd guess we want to do this across tables - the user probably doesn't care if we store something in the properties blob if it's useful for them to filter by

Was going to add documentation to the API docs. Is there a better place?

Comment (hazzadous) [reply ee/clickhouse/models/property.py:419]: API docs makes sense. I would also put something right here to clarify as if I'm making a change here, this is really were I would like the context. I can just imagine someone coming to the code and thinking why not just do this for all tables, removing the if, luckily the tests will catch this now but will be a big feedback cycle.

File: ee/clickhouse/models/property.py (hunk 3 above)
Comment (pauldambra) [reply ee/clickhouse/models/property.py:236]: Doh, great catch... Added test case to force that out

**Comment by pauldambra** [comment]
separated and (hopefully) clarified the tests @hazzadous 

Everything from the original PR is in this one

**Comment by hazzadous** [comment]
> Everything from the original PR is in this one

cool yup, I mean to say that I only looked at new commits

**Review by hazzadous — APPROVED** [review]
The updates look good, I haven't reviewed the older commits however, let me know if you want that as wel

File: posthog/api/property_definition.py [hunk 5]
```diff
@@ -104,23 +110,39 @@ def get_queryset(self):
...
+                   ('4771d3f7-f53d-4f35-96ee-cfb144af3e5e'::uuid,'distinct_id', false, null::int, null::int, 1, 'String', null::varchar, '4771d3f7-f53d-4f35-96ee-cfb144af3e5e'::uuid, '', null::varchar[], '1970-01-01 00:00+00'::timestamptz, 0, null::boolean)
+                )
                 SELECT posthog_propertydefinition.*,
                        ee_enterprisepropertydefinition.*, 
                        {event_property_field} AS is_event_property
                 FROM posthog_propertydefinition
                 LEFT JOIN ee_enterprisepropertydefinition ON ee_enterprisepropertydefinition.propertydefinition_ptr_id=posthog_propertydefinition.id
                 WHERE posthog_propertydefinition.team_id = %(team_id)s AND name NOT IN %(excluded_properties)s {name_filter} {numerical_filter} {search_query}
                 GROUP BY posthog_propertydefinition.id, ee_enterprisepropertydefinition.propertydefinition_ptr_id
+                -- the two filters start with AND, so 1=1 to avoid editing them
```
Comment (mariusandra) [inline posthog/api/property_definition.py:126]: neat!

**Review by mariusandra — APPROVED** [review]
Let's go for it

**Pull request merged**